copilot-bootstrap.py

```diff
@@ -409,76 +391,144 @@ def instructions(config_file):
+        conf.update(env_conf)
+
+        return conf
+
+    def _normalise_keys(source: dict):
+        return {k.replace("-", "_"): v for k, v in source.items()}
+
+    with open(BASE_DIR / "storage-plans.yaml", "r") as fd:
+        storage_plans = yaml.safe_load(fd)
+
+    with open(BASE_DIR / "schemas/storage-schema.json", "r") as fd:
+        schema = json.load(fd)
+
+    # load and validate config
+    with open(config_file, "r") as fd:
+        config = yaml.safe_load(fd)
+
+    validate_json(instance=config, schema=schema)
+
+    env_names = [path.parent.parts[-1] for path in Path("./copilot/environments/").glob("*/manifest.yml")]
+    svc_names = [path.parent.parts[-1] for path in Path("./copilot/").glob("*/manifest.yml")]
+
+    if not env_names:
+        click.echo(click.style(f"No environments found in ./copilot/environments; exiting", fg="red"))
+        exit(1)
+
+    if not svc_names:
+        click.echo(click.style(f"No services found in ./copilot/; exiting", fg="red"))
+        exit(1)
+
+    normalised_config = {}
+    for storage_name, storage_config in config.items():
+        storage_type = storage_config["type"]
+        normalised_config[storage_name] = copy.deepcopy(storage_config)
+
+        if "services" in normalised_config[storage_name]:
+            valid_services = [svc for svc in normalised_config[storage_name]["services"] if svc in svc_names]
+            if valid_services != normalised_config[storage_name]["services"]:
+                normalised_config[storage_name]["services"] = valid_services
+                click.echo(click.style(f"Services listed in {storage_name} do not exist in ./copilot/", fg="red"))
+
+        environments = normalised_config[storage_name].pop("environments", {})
+        default = environments.pop("default", {})
+
+        initial = _lookup_plan(storage_type, default)
+
+        normalised_environments = {}
+
+        for env in env_names:
+            normalised_environments[env] = _normalise_keys(initial)
+
+        for env_name, env_config in environments.items():
+            if env_name not in normalised_environments:
+                click.echo(click.style(f"Environment key {env_name} listed in {storage_name} does not exist in ./copilot/environments", fg="red"))
+            else:
+                normalised_environments[env_name].update(
+                    _lookup_plan(storage_type, _normalise_keys(env_config))
+                )
+
+        normalised_config[storage_name]["environments"] = normalised_environments
+
+    return normalised_config
+
+
 @cli.command()
 @click.argument("storage-config-file", type=click.Path(exists=True))
 @click.argument("output", type=click.Path(exists=True), default=".")
 @click.option('--overwrite', is_flag=True, show_default=True, default=True, help='Overwrite existing cloudformation? Defaults to True')
-def generate_storage(storage_config_file, output, overwrite):
+def make_cloudformation(storage_config_file, output, overwrite):
```

Comment:
Should we be specific about CloudFormation whether it's a stack, template, etc?

Reply:
Good question. I wonder what is the most descriptive from a developer perspective? We expect them to run this command anytime they make changes to `storage.yaml`, and anytime they add a new service or environment. 

'make-config' seems a bit too vague

@WillGibson what do you think?